h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java

```diff
@@ -0,0 +1,193 @@
+import water.fvec.FileVec;
+import water.fvec.GcsFileVec;
+import water.util.Log;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.util.*;
+
+/**
+ * Persistence backend for GCS
+ */
+public final class PersistGcs extends Persist {
+
+  private static final Storage storage = StorageOptions.getDefaultInstance().getService();
+
+  @Override
+  public byte[] load(final Value v) throws IOException {
+    final BlobId blobId = GcsBlob.of(v._key).getBlobId();
+    
+    final byte[] contentBytes = MemoryManager.malloc1(v._max);
+    final ByteBuffer wrappingBuffer = ByteBuffer.wrap(contentBytes);
+    final Key k = v._key;
+    long offset = 0;
+    // Skip offset based on chunk number
+    if(k._kb[0] == Key.CHK) {
+      offset = FileVec.chunkOffset(k); // The offset
+    }
+    
+    final ReadChannel reader = storage.reader(blobId);
+    reader.seek(offset);
+    reader.read(wrappingBuffer);
+    
+    return contentBytes;
+  }
+
+  @Override
+  public Key uriToKey(URI uri) throws IOException {
+    final GcsBlob blob = GcsBlob.of(uri);
+    final Long contentSize = storage.get(blob.getBlobId()).getSize();
+    return GcsFileVec.make(blob.getCanonical(), contentSize);
+  }
+
+  @Override
+  public void store(Value v) throws IOException {
+    if (!v._key.home()) return;
+    final byte payload[] = v.memOrLoad();
+    final GcsBlob blob = GcsBlob.of(v._key);
+    Log.debug("Storing: " + blob.toString());
+    final ByteBuffer buffer = ByteBuffer.wrap(payload);
+    storage.create(blob.getBlobInfo()).writer().write(buffer);
+  }
+
+  @Override
+  public void delete(Value v) {
+    final BlobId blobId = GcsBlob.of(v._key).getBlobId();
+    Log.debug("Deleting: " + blobId.toString());
+    storage.get(blobId).delete();
+  }
+
+  @Override
+  public void cleanUp() {
+    throw H2O.unimpl();
+  }
+
+  private static final Map<String, StringCache> keyCache = new HashMap<>();
+  private static final StringCache bucketCache = new StringCache() {
+    @Override
+    List<String> update() {
+      final List<String> c = new ArrayList<>();
+      for (Bucket b : storage.list().iterateAll()) {
+        c.add(b.getName());
+      }
+      return c;
+    }
+  };
+
+  @Override
+  public List<String> calcTypeaheadMatches(String filter, int limit) {
+    final String input = GcsBlob.removePrefix(filter);
+    final String[] bk = input.split("/", 2);
+    if (bk.length == 1) {
+      return bucketCache.fetch(GcsBlob.KEY_PREFIX, bk[0], limit);
+    } else if (bk.length == 2) {
+      if (!keyCache.containsKey(bk[0])) {
+        StringCache sc = new StringCache() {
+          @Override
+          List<String> update() {
+            final List<String> cache = new ArrayList<>();
+            for (Blob b : storage.get(bk[0]).list().iterateAll()) {
+              cache.add(b.getName());
+            }
+            return cache;
+          }
+        };
+        keyCache.put(bk[0], sc);
```

Comment:
I would avoid concurrent modifications of the non-threadsafe hashmap

Reply:
Resolved by using Guava's solution.